lib/client.c

```diff
@@ -1992,50 +1939,135 @@ SF_STATUS STDCALL _snowflake_execute_ex(SF_STMT *sfstmt,
                         SET_SNOWFLAKE_STMT_ERROR(&sfstmt->error,
-                                                 SF_STATUS_ERROR_BAD_JSON,
-                                                 "Missing rowset from response. No results found.",
-                                                 SF_SQLSTATE_APP_REJECT_CONNECTION,
-                                                 sfstmt->sfqid);
+                            SF_STATUS_ERROR_BAD_JSON,
+                            "Missing rowset from response. No results found.",
+                            SF_SQLSTATE_APP_REJECT_CONNECTION,
+                            sfstmt->sfqid);
+                        goto cleanup;
+                    }
+                }
+                else if (strcmp(qrf_str, "json") == 0) {
+                    *((QueryResultFormat_t *) sfstmt->qrf) = JSON_FORMAT;
+                    if (json_detach_array_from_object((cJSON **)(&rowset), data, "rowset"))
+                    {
+                        log_error("No valid rowset found in response");
+                        SET_SNOWFLAKE_STMT_ERROR(&sfstmt->error,
+                            SF_STATUS_ERROR_BAD_JSON,
+                            "Missing rowset from response. No results found.",
+                            SF_SQLSTATE_APP_REJECT_CONNECTION,
+                            sfstmt->sfqid);
+                        goto cleanup;
+                    }
+                }
+                else {
+                    log_error("Unsupported query result format: %s", qrf_str);
+                }
+
+                // Index starts at 0 and incremented each fetch
+                sfstmt->total_row_index = 0;
+
+                // When the result set is sufficient large, the server response will contain
+                // an empty "rowset" object. Instead, it will have a "chunks" object that contains,
+                // among other fields, a URL from which the result set can be downloaded in chunks.
+                // In this case, we initialize the chunk downloader, which will download in the
+                // background as calls to snowflake_fetch() are made.
+                if ((chunks = snowflake_cJSON_GetObjectItem(data, "chunks")) != NULL) {
+                    // We don't care if there is no qrmk, so ignore return code
+                    json_copy_string(&qrmk, data, "qrmk");
+                    chunk_headers = snowflake_cJSON_GetObjectItem(data, "chunkHeaders");
+                    NON_JSON_RESP* (*callback_create_resp)(void) = NULL;
+                    if (ARROW_FORMAT == *((QueryResultFormat_t *)sfstmt->qrf)) {
+                        callback_create_resp = callback_create_arrow_resp;
+                    }
+
+                    sfstmt->chunk_downloader = chunk_downloader_init(
+                            qrmk,
+                            chunk_headers,
+                            chunks,
+                            2, // thread count
```

Comment:
Should these number conquered by parameter client_prefetch_threads?

Reply:
@sfc-gh-cshi 
That's the original implementation and unchanged. https://github.com/snowflakedb/libsnowflakeclient/blob/master/lib/client.c#L2032
I think client_prefetch_threads is not supported by libsnowflakeclient before. If that's needed, it would be another task and made in a different PR.